modules/packages/Sort.chpl

```diff
@@ -18,53 +18,321 @@
+    throw a compile-time error.
+
+    :arg a: Sample data passed to confirm that comparator methods can resolve
+    :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+private proc chpl_check_comparator(comparator, type eltType) {
+  use Reflection;
+
+  // Dummy data for checking method resolution
+  // This may need updating when constructors support non-default args
+  const data: eltType;
+
+  if comparator.type == DefaultComparator {}
+  // Check for valid comaparator methods
+  else if canResolveMethod(comparator, "compare", data, data) {
+    // Check return type of compare
+    type comparetype = comparator.compare(data, data).type;
+    if !(isNumericType(comparetype)) {
+      compilerError("The compare method must return a numeric type");
+    }
+  }
+  else if canResolveMethod(comparator, "key", data) {
+    // Check return type of key
+    type keytype = comparator.key(data).type;
+    if !(isNumericType(keytype) || isStringType(keytype)) {
+      compilerError("The key method must return a numeric or string type");
+    }
+  }
+  else {
+    // If we make it this far, the passed comparator was defined incorrectly
+    compilerError("The comparator record requires a 'key(a)' or 'compare(a, b)' method");
+  }
+}
+
+
+/* Basic Functions */
+
+/*
+   Convenience wrapper around :proc:`quickSort`.
+
+   :arg Data: The array to be sorted
+   :type Data: [] `eltType`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+proc sort(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator) where Dom.rank == 1 {
+  quickSort(Data, comparator=comparator);
+}
+
+
+/*
+   Check if array `Data` is in sorted order
+
+   :arg Data: The array to verify
+   :type Data: [] `eltType`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+   :returns: ``true`` if array is sorted
+   :rtype: `bool`
+ */
+proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator): bool {
+  chpl_check_comparator(comparator, eltType);
+  for i in Dom.low..Dom.high-1 do
+    if chpl_compare(Data(i+1), Data(i), comparator) < 0 then
+      return false;
+  return true;
+}
+
+
+//
+// This is a first draft "sorterator" which is designed to take some
+// other iterator/iterable and yield its elements, in sorted order.
+//
+// The main limitations in the current code are (1) it should put some
+// sort of constraint on 'x' to limit it to types for which this makes
+// sense; and (2) there should be some generic way to say "y is an
+// array of x's element type" (or to infer its element type) without
+// saying a priori how big it is.  Without these mods, the result is
+// that the sorterator works when it does and probably is confusing
+// when it doesn't.
```

Comment:
Original comment by @bradcray, so he can correct me if I'm off:

I think it is saying that it would preferably create a new array `y` which shares the same element types `x`, but this requires a size to be passed (or we could calculate the size on the fly). This would be valuable for supporting iterable types other than arrays, such as iterators and lists.